### PR TITLE
feat(cli): improve CLI transient queries with headers and spacing (partial fix for #892)

### PR DESCRIPTION
**Pull request by agavra**
### Description 
Ladies and Gentlemen, the moment you have _all_ been waiting for! Hot off the press 🔥 🔥 🔥 Formatted CLI output for transient queries, with a shiny new header row 🎉 🎉 🎉 

```sql
ksql> SELECT * FROM ALL_PUBLICATIONS;
+-----------------------+-----------------------+-----------------------+-----------------------+
|ROWTIME                |ROWKEY                 |AUTHOR                 |TITLE                  |
+-----------------------+-----------------------+-----------------------+-----------------------+
|1562173641326          |C.S. Lewis             |C.S. Lewis             |The Silver Chair       |
|1562173641364          |George R. R. Martin    |George R. R. Martin    |A Song of Ice and Fire |
|1562173641378          |C.S. Lewis             |C.S. Lewis             |Perelandra             |
|1562173641392          |George R. R. Martin    |George R. R. Martin    |Fire & Blood           |
|1562173641406          |J. R. R. Tolkien       |J. R. R. Tolkien       |The Hobbit             |
|1562173641419          |J. R. R. Tolkien       |J. R. R. Tolkien       |The Lord of the Rings  |
|1562173641432          |George R. R. Martin    |George R. R. Martin    |A Dream of Spring      |
|1562173641446          |J. R. R. Tolkien       |J. R. R. Tolkien       |The Fellowship of the R|
|                       |                       |                       |ing                    |
|1562173641460          |George R. R. Martin    |George R. R. Martin    |The Ice Dragon         |
```

It even does very fancy 💃 🕺 wrapping for lines that are too long and adjusts based on your terminal size.

### Notes

This has a hacked version of https://github.com/confluentinc/ksql/pull/3043 as part of it, we should resolve that independently

### Testing done 

- Unit testing
- CLI integration testing

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

**Comment by MichaelDrogalis** [comment]
May I suggest a slight cosmetic adjustment to match MySQL?

```
+---------+--------+-------+
| article | dealer | price |
+---------+--------+-------+
|       1 | A      |  3.45 |
|       1 | B      |  3.99 |
|       2 | A      | 10.99 |
|       3 | B      |  1.45 |
|       3 | C      |  1.69 |
|       3 | D      |  1.25 |
|       4 | D      | 19.95 |
+---------+--------+-------+
```

- Top row of dashes before the headers
- Row of dashes under the headers
- No dashes between rows

Mostly curious to see how it looks if we try to line 'em up

**Comment by hjafarpour** [comment]
This is great @agavra , I agree with @MichaelDrogalis 's comment on matching MySQL. Postgres also is very similar to MySQL. That way it would really look and feel like a database! ;)

File: ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java [hunk 1]
```diff
@@ -323,6 +325,18 @@ private void printKsqlResponse(final RestResponse<KsqlEntityList> response) thro
 
   @SuppressWarnings("try")
   private void handleStreamedQuery(final String query) throws IOException {
+    final RestResponse<KsqlEntityList> explainResponse = restClient
+        .makeKsqlRequest("EXPLAIN " + query);
+    final List<FieldInfo> fields;
+    if (!explainResponse.isSuccessful()) {
+      terminal.printErrorMessage(explainResponse.getErrorMessage());
+      return;
+    } else {
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/cli/Cli.java]: nit: unnecessary else, as if returns.

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class TabularRow {
+
+  private final int width;
+  private final List<String> value;
+  private final List<String> header;
+
+  public TabularRow(
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: `header` and `value` are used mutually exclusively.   They both map to a list of strings.  So this class could be simplified to just take the width and single `List<String>` param, pushing up the computation of _what_ is passed as the list of strings to the caller, or maybe two factory methods.

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+          Splitter.fixedLength(cellWidth)
+              .splitToList(column)
+              .stream()
+              .map(line -> String.format(format, line))
+              .collect(Collectors.toList())
+      );
+    }
+
+    // buffer each column vertically to have the max number of splits
+    final int maxSplit = split.stream().map(List::size).max(Integer::compareTo).orElse(0);
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]:
```suggestion
    final int maxSplit = split.stream().mapToInt(List::size).max().orElse(0);
```

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+      final List<String> header,
+      final List<String> value
+  ) {
+    this.header = Objects.requireNonNull(header, "header");
+    this.width = width;
+    this.value = value;
+  }
+
+  @Override
+  public String toString() {
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: this method would benefit from being split into several methods, rather than being long with comments.

Something like:

```java
public String toString() {
final List<String> columns = value == null ? header : value;

     final int cellWidth = width / columns.size() - 2;
    final String format = "%-" + cellWidth + "s|";

    final List<List<String>> split = fitToCellWidth();

    final int maxSplit = maxLines(split);

    return formatRow(split, cellWidth, ...)
       + buildDiv(coluns.size(), cellWidth);
}
```

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlBareOutputNode.java [hunk 5]
```diff
@@ -39,7 +39,17 @@ public KsqlBareOutputNode(
       final OptionalInt limit,
       final TimestampExtractionPolicy extractionPolicy
   ) {
-    super(id, source, schema, limit, extractionPolicy);
+    super(
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlBareOutputNode.java]: Don't think adding this was intentional....  we shouldn't be doing this.

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/EntityUtil.java [hunk 6]
```diff
@@ -42,10 +42,6 @@ private EntityUtil() {
   }
 
   private static List<FieldInfo> getFields(final List<Field> fields, final String type) {
-    if (fields.isEmpty()) {
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/EntityUtil.java]: Likewise, I don't think we should be doing this either...

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java [hunk 7]
```diff
@@ -285,6 +285,15 @@ public LogicalSchema withoutMetaAndKeyFieldsInValue() {
         alias);
   }
 
+  public LogicalSchema withoutMetaAndKeyFields() {
```
Comment (big-andy-coates) [inline ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java]: We shouldn't be doing this.... I've a fix for the bug here: https://github.com/confluentinc/ksql/pull/3043

File: ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    final String formatted = new TabularRow(20, header, header).toString();
+
+    // Then:
+    assertThat(formatted, is(""
+        + "|foo     |bar     |\n"
+        + "+--------+--------+"));
+  }
+
+  @Test
+  public void shouldMultilineFormatRow() {
```
Comment (big-andy-coates) [inline ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java]: Add test for what happens if the number of columns is higher than the width....

Interesting edge case!

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java (hunk 2 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: the reason I have it this way is that in the near future I want to change the column format based on the header - so passing in both will soon be helpful (e.g. fixed width for int columns, etc...)

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java (hunk 3 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: nice!

File: ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlBareOutputNode.java (hunk 5 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/planner/plan/KsqlBareOutputNode.java]: I was just hacking it to make it work - wasn't going to commit until your fix is in :)

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/EntityUtil.java (hunk 6 above)
Comment (agavra) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/util/EntityUtil.java]: Same as above

File: ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java (hunk 7 above)
Comment (agavra) [reply ksql-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java]: Ditto

File: ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java (hunk 8 above)
Comment (agavra) [reply ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java]: 😨 good catch! I think if this is the case I'll default back to the old behavior

**Comment by spena** [comment]
@agavra While testing this, I see that the output fits the full screen. Can we use a fixed column instead? Mysql and other Dbs do not fit the whole screen, although I understand that those DBs know what's the max value read to use it in the column width, and KSQL does not know as it is streaming real-time data.

Btw, if I stop the query and rerun it again, the headers are not displayed anymore.
```
ksql> select * from t1;
|ROWTIME                                                |ROWKEY                                                 |ID                                                     |
+═══════════════════════════════════════════════════════+═══════════════════════════════════════════════════════+═══════════════════════════════════════════════════════+
|1562104196553                                          |null                                                   |1                                                      |
+-------------------------------------------------------+-------------------------------------------------------+-------------------------------------------------------+
^CQuery terminated
      select * from t1;
|1562104196553                                                                     |null                                                                              |1                                                                                 |
+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+
ksql> select * from t1;
|1562104196553                                                                     |null                                                                              |1                                                                                 |
+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+
ksql> 
ksql> 
ksql> select * from t1;
|1562104196553                                                                     |null                                                                              |1                                                                                 |
+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+----------------------------------------------------------------------------------+
```

**Comment by agavra** [comment]
@spena - I'm not sure what's going on with your terminal, but my headers do show multiple times... 

and yeah, my concern was that I don't know the max value length when I start it up. The other SQL engines also have more powerful formatting options for the user to choose how they want it to show up. I think until we implement those, we should have a good default option (which, to me, full width makes sense)

File: ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java (hunk 8 above)
Comment (agavra) [reply ksql-cli/src/test/java/io/confluent/ksql/util/TabularRowTest.java]: I've decided to set a minimum cell width of 5 - it'll look all weird if the terminal is super small, but that's going to be the case anyway

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java (hunk 2 above)
Comment (big-andy-coates) [reply ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: OK, in that case I'd still go for two factory methods to make this clear:

`TabularRow.createHeader(headers)` vs `TabularRow.createRow(headers, row)`.

And you can still decouple this class from `FieldInfo` and `GenericRow` by having the calling code do the `toString` stuff.  Less couple is good!  Plus, you can process the headers once for the query, rather than per-row.

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    if (isHeader) {
+      builder.append('\n');
+      separatingLine(builder, cellWidth, columns.size());
+    }
+
+    return builder.toString();
+  }
+
+  @SuppressWarnings("ForLoopReplaceableByForEach") // clearer to read this way
+  private static void toString(
```
Comment (big-andy-coates) [inline ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: nit: rename `formatRow` or something.

File: ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java (hunk 2 above)
Comment (agavra) [reply ksql-cli/src/main/java/io/confluent/ksql/util/TabularRow.java]: > And you can still decouple this class from FieldInfo and GenericRow by having the calling code do the toString stuff.

I don't think the _current_ calling class would be a good place to do that. I could introduce another wrapper class (e.g. `Tabular`) that handled all of the row widths based on the field infos and things like that, but that is overkill for this PR. As it stands, it's pretty decoupled - it's just in a constructor (moved to be static initializer).
